### PR TITLE
Added raw file upload page including tests.

### PR DESCRIPTION
**Pull request by Donnype**
Signed-off-by: Donny Peeters <donny@bitestreams.com>

## Changes
Much like the Upload CSV page, this PR adds an Upload raw page where you specify a file and can add mime types.

## Issue ticket number and link
https://github.com/minvws/nl-kat-coordination/issues/158

## Proof
image

## Good to know
- Does this PR introduce or depend on API-incompatible changes? If yes: what do other users/developers need to do or confirm before merging?
  - (Yes/No here, plus explanation if necessary)
- Does this PR depend on a specific version of a library?
  - (Yes/No here, plus explanation if necessary)
- Does this PR require config, setup, or `.env` changes?
  - (Yes/No here, plus explanation if necessary)

## Checklist for author(s):
- [x] This PR comes from a `feature` or `hotfix` branch, in line with our git branching strategy;
- [x] I have squashed all commits before merging this PR;
- [x] This PR is "bite-sized" and only focuses on a single issue, problem, or feature;
- [x] If a non-trivial PR: This PR is properly linked to an issue on the project board;
- [x] If a non-trivial PR: I have added screenshots or some other proof that my code does what it is supposed to do;
- [x] I am not reinventing the wheel: there is no high-quality library that already has this feature;
- [x] I have changed the example `.env` files if I added, removed, or changed any config options, and I have informed others that they need to modify their `.env` files if required;
- [x] I have performed a self-review of my own code;
- [x] I have commented my code, particularly in hard-to-understand areas;
- [x] I have made corresponding changes to the documentation, if necessary;
- [x] I have written unit, integration, and end-to-end tests for the change that I made;

```
## Checklist for functional reviewer(s):
- [ ] If a non-trivial PR: This PR is properly linked to an issue on the project board;
- [ ] I have checked out this branch, and successfully ran `make kat`;
- [ ] I have ran `make test-rf` and all end-to-end Robot Framework tests pass;
- [ ] I confirmed that the PR's advertised `feature` or `hotfix` works as intended;
- [ ] I confirmed that there are no unintended functional regressions in this branch;

### What works:
* _bullet point + screenshot (if useful) per tested functionality_

### What doesn't work:
* _bullet point + screenshot (if useful) per tested functionality_

### Bug or feature?:
* _bullet point + screenshot (if useful) if it is unclear whether something is a bug or an intended feature._
```

```
## Checklist for code reviewer(s):
- [ ] The code passes the CI tests and linters;
- [ ] The code does not bypass authentication or security mechanisms;
- [ ] The code does not introduce any dependency on a library that has not been properly vetted;
- [ ] The code does not violate Model-View-Template and our other architectural principles;
- [ ] The code contains docstrings, comments, and documentation where needed;
- [ ] The code prioritizes readability over performance where appropriate;
- [ ] The code conforms to our agreed coding standards.
```

### CODE REVIEW

File: tools/forms/upload_raw.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    raw_file = forms.FileField(
+        label=_("Upload raw file"),
+        allow_empty_file=False,
+    )
+
+    def clean_raw_file(self):
+        raw_file = self.cleaned_data["raw_file"]
+
+        try:
+            raw_file.read().decode("UTF-8")
```
Comment (ammar92) [inline tools/forms/upload_raw.py:37]: Do we assume that raw files are always (text) files encoded with UTF8 encoding? This way you can't upload any file format as suggested by the MIME type field

**Review by ammar92 — COMMENTED** [review]
Nice stuff. I just had one question about the file input format and wanted to be sure that it isn't a mistake. Consider it approved otherwise

**Comment by Darwinkel** [comment]
## Checklist for functional reviewer(s):
- [x] If a non-trivial PR: This PR is properly linked to an issue on the project board;
- [x] I have checked out this branch, and successfully ran `make kat`;
- [ ] I have ran `make test-rf` and all end-to-end Robot Framework tests pass;
- [x] I confirmed that the PR's advertised `feature` or `hotfix` works as intended;
- [x] I confirmed that there are no unintended functional regressions in this branch;

### What works:
* A raw HTML file is uploaded correctly to bytes, and is parsed.
image
* A raw Calvin file is uploaded correctly to bytes, and is parsed. New findings are correctly created afterwards.

### What doesn't work:
* No new finding or object is created after the normalizer finishes parsing the HTML - `input_ooi` is missing.

### Bug or feature?:
* Currently, the page does not describe how raw files can be obtained or created. I think it would be very helpful if the page contained some concrete examples. It could also mention that more examples can be found by going to the Tasks->Boefjes or Tasks->Normalizers page, and then downloading a zip containing the raw data.

File: tools/forms/upload_raw.py (hunk 1 above)
Comment (Donnype) [reply tools/forms/upload_raw.py:37]: Ah fair point

Comment (Donnype) [reply tools/forms/upload_raw.py]: Fixed

**Review by ammar92 — APPROVED** [review]
No remarks!

**Comment by Donnype** [comment]
I think the issues @Darwinkel found are a fair point: do we want to be able to optionally link an `input_ooi` when uploading a raw file or do are we going to restrict this feature to normalizers creating declarations? And: are we going to assume users needing/using this page often built the normalizer they want to trigger themselves, so they know what the raw file should look like? CC @underdarknl

**Comment by Donnype** [comment]
Refinement ticket: https://github.com/minvws/nl-kat-coordination/issues/173

**Review by Darwinkel — APPROVED** [review]
Keeping the refinement ticket in mind for the next sprint, this PR is OK.

**Comment by Donnype** [comment]
Closing because of migration to https://github.com/minvws/nl-kat-coordination/

New PR:
https://github.com/minvws/nl-kat-coordination/pull/175

**Pull request closed without merging**